File: PD2/testB.java
public class Main {
  static int MASTER_NODE = 0;
  static long SENDING_DONE = -1;
  static long LARGE_PRIME = 1000000007;

  public static void main(String[] args) {
    if (message.MyNodeId() == MASTER_NODE) {
      long result = 0;
      for (int node = 1; node < message.NumberOfNodes(); ++node) {
        while (true) {
          message.Receive(node);
          long value = message.GetLL(node);
          if (value == SENDING_DONE) {
            break;
          } else {
            result = (result + value) % LARGE_PRIME;
          }
        }
      }
      System.out.println(result);
    } else {
      long tmpresult = 0L;
      int n = (int) again.GetN();
      long[] arra = new long[n];
      long[] arrb = new long[n];
      for (int i = 0; i < n; ++i) {
        arra[i] = again.GetA((long) i);
        arrb[i] = again.GetB((long) i);
      }
      for (int i = 0; i < n; ++n) {
        for (int j = 0; j < n; ++j) {
          if ((i + j) % message.NumberOfNodes() == message.MyNodeId()) {
            long value = arra[i] * arrb[j];
            tmpresult = (tmpresult + value) % LARGE_PRIME;
          }
        }
      }
      message.PutLL(MASTER_NODE, tmpresult);
      message.Send(MASTER_NODE);
      message.PutLL(MASTER_NODE, SENDING_DONE);
      message.Send(MASTER_NODE);
    }
  }
}

File: PD2/testB.cpp
#include <cstdio>
#include "again.h"
#include "message.h"

#define MAX_NODES 20
#define MOD 1000000007

using namespace std;

typedef long long ll;

ll a[MAX_NODES], b[MAX_NODES];

int main() {
  int nodeNum = NumberOfNodes();
  int nodeBegin = (int) (MyNodeId() * GetN() / nodeNum);
  int nodeEnd = (int) ((MyNodeId() + 1) * GetN() / nodeNum);

  for(int i = nodeBegin; i < nodeEnd; i++) {
    int k = i % nodeNum;
    a[k] = (a[k] + GetA(i)) % MOD;
    b[k] = (b[k] + GetB(i)) % MOD;
  }

  if(MyNodeId() != 0) {
    for(int k = 0; k < nodeNum; k++) {
      PutLL(0, a[k]);
      PutLL(0, b[k]);
    }
    Send(0);
  } else {
    for(int k = 1; k < nodeNum; k++) {
      Receive(k);
      for(int i = 0; i < nodeNum; i++) {
        a[i] = (a[i] + GetLL(k)) % MOD;
        b[i] = (b[i] + GetLL(k)) % MOD;
      }
    }

    ll res = 0;
    for(int i = 0; i < nodeNum; i++) {
      for(int j = 0; j < nodeNum; j++) {
        if((i + j) % nodeNum > 0) {
          res = (res + (a[i] * b[j] % MOD)) % MOD;
        }
      }
    }
    printf("%lld\n", res);
  }
  return 0;
}